### PR TITLE
feat(isthmus): observe scalar function types

### PR DESCRIPTION
**Pull request by alexandrefimov**
## Summary

- add an opt-in type observer for Substrait-to-Calcite scalar function conversion
- report the source expression, supplied Substrait type, and either the independently inferred Calcite type or the inference failure
- expose observer configuration through `ConverterProvider` while keeping the default conversion path unchanged

Related to #379.

## Behavior

- observations are emitted for both matching and differing scalar function types
- the converted `RexNode` continues to use the type supplied by Substrait
- Calcite inference failures are reported without turning an otherwise successful conversion into a failure
- exceptions thrown by the observer are propagated to the caller
- variadic `CONCAT` produces one observation for the complete invocation

This PR intentionally covers scalar functions only. Window functions, field references, and aggregate functions remain separate follow-up work.

## Validation

- `./gradlew build --rerun-tasks`
- focused scalar observation tests covering matches, mismatches, inference failures, nested calls, provider injection, observer exceptions, and variadic `CONCAT`

Assisted-by: gpt-5.6-sol (OpenAI Codex)

### CODE REVIEW

**Comment by nielspardon** [comment]
@vbarua can you take a look since you originally opened #379?

**Comment by vbarua** [comment]
Did a quick pass and the direciton seems reasonable overall. Planning on taking a closer look tomorrow.

File: isthmus/src/main/java/io/substrait/isthmus/expression/ExpressionRexConverter.java [hunk 1]
```diff
@@ -522,13 +551,49 @@ public RexNode visit(Expression.ScalarFunctionInvocation expr, Context context)
...
+      observeScalarType(
+          expr,
+          () ->
+              args.stream()
+                  .skip(1)
+                  .reduce(
+                      args.get(0),
+                      (left, right) -> rexBuilder.makeCall(operator, List.of(left, right))));
+      return suppliedCall;
+    }
```
Comment (vbarua) [inline isthmus/src/main/java/io/substrait/isthmus/expression/ExpressionRexConverter.java]: I got a little nerdsniped seeing that you had to work around this special case CONCAT handling, and am moving to just fully remove the handling which is both cleaner overall for dealing with CONCAT, and should simplify your work here.

https://github.com/substrait-io/substrait-java/pull/1025

File: isthmus/src/main/java/io/substrait/isthmus/expression/ExpressionRexConverter.java [hunk 2]
```diff
@@ -522,13 +551,49 @@ public RexNode visit(Expression.ScalarFunctionInvocation expr, Context context)
...
+    RexNode suppliedCall = rexBuilder.makeCall(returnType, operator, args);
```
Comment (vbarua) [inline isthmus/src/main/java/io/substrait/isthmus/expression/ExpressionRexConverter.java]: minor: I would just suggest calling this something like `rexCallWithForcedSubstraitReturnType`, or just `rexCall`.

File: isthmus/src/main/java/io/substrait/isthmus/expression/ExpressionRexConverter.java [hunk 3]
```diff
@@ -522,13 +551,49 @@ public RexNode visit(Expression.ScalarFunctionInvocation expr, Context context)
...
+    if (typeObserver == TypeObserver.NOOP) {
```
Comment (vbarua) [inline isthmus/src/main/java/io/substrait/isthmus/expression/ExpressionRexConverter.java:554]: A comment to the effect of
"If Type Observations are not needed, we return the call as is an avoid the work to recompute a RexCall with Calcite's inferred return type."
would be helpful here to explain why we're short-circuiting, and what we're doing below.

File: isthmus/src/main/java/io/substrait/isthmus/expression/ExpressionRexConverter.java [hunk 4]
```diff
@@ -522,13 +551,49 @@ public RexNode visit(Expression.ScalarFunctionInvocation expr, Context context)
...
+
+  private void observeScalarType(
+      Expression.ScalarFunctionInvocation expression, Supplier<RexNode> inferredCallSupplier) {
+    TypeObservation observation;
+    RexNode inferredCall;
+    try {
+      inferredCall = inferredCallSupplier.get();
+    } catch (RuntimeException exception) {
+      observation =
+          TypeObservation.failure(TypeObservation.Source.SCALAR_FUNCTION, expression, exception);
```
Comment (vbarua) [inline isthmus/src/main/java/io/substrait/isthmus/expression/ExpressionRexConverter.java]: minor: I might suggest `failure` -> `inferenceFailure` or `calciteInferenceFailure` to make it clearer what failed. It took me a bit to realize we weren't checking for type differences here (which makes sense, that will come later).

I don't feel super strongly about it.

File: isthmus/src/main/java/io/substrait/isthmus/expression/TypeObservation.java [hunk 5]
```diff
@@ -0,0 +1,109 @@
+package io.substrait.isthmus.expression;
+
+import io.substrait.expression.Expression;
+import io.substrait.type.Type;
+import java.util.Objects;
+import java.util.Optional;
+import org.apache.calcite.rel.type.RelDataType;
+
+/**
+ * The result of attempting to independently infer a Calcite type during expression conversion.
```
Comment (vbarua) [inline isthmus/src/main/java/io/substrait/isthmus/expression/TypeObservation.java]: I would specify Substrait to Calcite expression conversion... though could we ever do TypeObservations in the opposite direction (not something to implement in this PR).

File: isthmus/src/test/java/io/substrait/isthmus/SubstraitExpressionConverterTest.java [hunk 6]
```diff
@@ -193,4 +361,47 @@ void assertTypeMatch(RelDataType actual, Type expected) {
     Type type = TypeConverter.DEFAULT.toSubstrait(actual);
     assertEquals(expected, type);
   }
+
+  private Expression.ScalarFunctionInvocation add(Type outputType) {
```
Comment (vbarua) [inline isthmus/src/test/java/io/substrait/isthmus/SubstraitExpressionConverterTest.java]: I would give this a more specific name like `integerAddWithReturnType`

**Review by vbarua — COMMENTED** [review]
Left some minor comments, but looks good overall.

**Comment by alexandrefimov** [comment]
Thanks for the review! The five minor comments are clear. I’ll address them together with the CONCAT simplification after #1025 is merged.

**Comment by vbarua** [comment]
CONCAT simplification is merged 🙂

File: isthmus/src/main/java/io/substrait/isthmus/expression/ExpressionRexConverter.java (hunk 1 above)
Comment (alexandrefimov) [reply isthmus/src/main/java/io/substrait/isthmus/expression/ExpressionRexConverter.java]: Rebased on `main` after #1025 and removed the special-case CONCAT handling. Variadic CONCAT still has coverage asserting a single observation for the complete invocation.

File: isthmus/src/main/java/io/substrait/isthmus/expression/ExpressionRexConverter.java (hunk 2 above)
Comment (alexandrefimov) [reply isthmus/src/main/java/io/substrait/isthmus/expression/ExpressionRexConverter.java]: Renamed this to `rexCall`.

File: isthmus/src/main/java/io/substrait/isthmus/expression/ExpressionRexConverter.java (hunk 3 above)
Comment (alexandrefimov) [reply isthmus/src/main/java/io/substrait/isthmus/expression/ExpressionRexConverter.java:554]: Added the short-circuit explanation. I also added a regression test asserting that the default NOOP observer performs zero Calcite inference calls.

File: isthmus/src/main/java/io/substrait/isthmus/expression/ExpressionRexConverter.java (hunk 4 above)
Comment (alexandrefimov) [reply isthmus/src/main/java/io/substrait/isthmus/expression/ExpressionRexConverter.java]: Renamed it to `inferenceFailure`.

File: isthmus/src/main/java/io/substrait/isthmus/expression/TypeObservation.java (hunk 5 above)
Comment (alexandrefimov) [reply isthmus/src/main/java/io/substrait/isthmus/expression/TypeObservation.java]: Updated the Javadoc to specify Substrait-to-Calcite expression conversion.

File: isthmus/src/test/java/io/substrait/isthmus/SubstraitExpressionConverterTest.java (hunk 6 above)
Comment (alexandrefimov) [reply isthmus/src/test/java/io/substrait/isthmus/SubstraitExpressionConverterTest.java]: Renamed the helper to `integerAddWithReturnType`.

**Comment by alexandrefimov** [comment]
@vbarua rebased on main and addressed all review comments. Ready for another look when you have a chance. Thanks!

**Comment by alexandrefimov** [comment]
@vbarua, gentle follow-up when you have a chance :)

**Review by vbarua — APPROVED** [review]
Changes look good. Thanks for work on this 🙇

**Comment by vbarua** [comment]
@alexandrefimov as a follow-up, if you wanted to make it easier to use the observer you've added, you could add it as a configurable setting in ConverterProvider.Builder
https://github.com/substrait-io/substrait-java/blob/243be42d193e305f6a6839735bbc337488c5aa0a/isthmus/src/main/java/io/substrait/isthmus/ConverterProvider.java#L546-L554

**Comment by alexandrefimov** [comment]
Thanks!
